geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/netty/Coder.java

```diff
@@ -471,4 +472,101 @@ public static int narrowLongToInt(long toBeNarrowed) {
     }
     return doubleBytes;
   }
+
+  /**
+   * Takes the given "value" and computes the sequence of ASCII digits it represents,
+   * appending them to the given "buf".
+   * This code was adapted from the openjdk Long.java getChars methods.
+   */
+  public static void appendAsciiDigitsToByteBuf(long value, ByteBuf buf) {
+    if (value < 0) {
+      buf.writeByte('-');
+    } else {
+      value = -value;
+    }
+    // at this point value <= 0
+
+    if (value > -100) {
+      // it has at most two digits: [0..99]
+      appendSmallAsciiDigitsToByteBuf((int) value, buf);
+    } else {
+      appendLargeAsciiDigitsToByteBuf(value, buf);
+    }
+  }
+
+  private static void appendSmallAsciiDigitsToByteBuf(int value, ByteBuf buf) {
+    int q = value / 10;
+    int r = (q * 10) - value;
+    if (q < 0) {
+      buf.writeByte((byte) ('0' - q));
+    }
+    buf.writeByte((byte) ('0' + r));
+  }
+
+  private static void appendLargeAsciiDigitsToByteBuf(long value, ByteBuf buf) {
+    long q;
+    int r;
+    final int MAX_DIGITS = 19;
+    int charPos = MAX_DIGITS;
+    byte[] bytes = new byte[MAX_DIGITS];
+
+    // Get 2 digits/iteration using longs until quotient fits into an int
+    while (value <= Integer.MIN_VALUE) {
+      q = value / 100;
+      r = (int) ((q * 100) - value);
+      value = q;
+      bytes[--charPos] = DIGIT_ONES[r];
+      bytes[--charPos] = DIGIT_TENS[r];
+    }
+
+    // Get 2 digits/iteration using ints
+    int q2;
+    int i2 = (int) value;
+    while (i2 <= -100) {
+      q2 = i2 / 100;
+      r = (q2 * 100) - i2;
+      i2 = q2;
+      bytes[--charPos] = DIGIT_ONES[r];
+      bytes[--charPos] = DIGIT_TENS[r];
+    }
+
+    // We know there are at most two digits left at this point.
+    q2 = i2 / 10;
+    r = (q2 * 10) - i2;
+    bytes[--charPos] = (byte) ('0' + r);
+
+    // Whatever left is the remaining digit.
+    if (q2 < 0) {
+      bytes[--charPos] = (byte) ('0' - q2);
+    }
+    buf.writeBytes(bytes, charPos, MAX_DIGITS - charPos);
+  }
+
+  @Immutable
+  private static final byte[] DIGIT_TENS = {
```

Comment:
If we want to follow the "important note" about not using chars mentioned above, then these arrays will need to use the UTF-8 byte values associated with these chars, which will probably make it look horrible.

Reply:
these static byte arrays came straight from the jdk so our old code was already using them when it called String.valueOf(long) which called Long.toString(long) which ended up in this code. Even the other code you commented on was for the most part from the jdk and was using '0' and '-'. So either the old code was already broken or the new code is okay.

Reply:
Maybe what made the old code safe is that after the jdk converted a long to a String we then called stringToBytes which converted to String chars to the hardcoded UTF-8 charset. In this new code that is no longer happening so if they are using a default charset that does not map '0' to the ASCII code for 0 then this code will be broken. I'll convert it over to using the constants.

Reply:
these static arrays are now initialized with some static code using the byte constants.